### PR TITLE
endpoints for create task nowait and getting event stream

### LINKED ISSUE #296: Add endpoint to API Server `deployments` resource to get event stream using client's `get_task_result_stream` method.
(The control plane already consumes the stream messages from the Workflow services and emits SSE using FastAPI StreamingResponse)

Implement the equivalent of what we have in workflows:

```python
handler = workflow.run()
async for event in handler.stream_events():
    # do something with event
    pass
```

for Llama Deploy through the API Server, something like GET http://apiserver.local/my-deployment/events

**Acceptance Criteria**
- there is an endpoint in the API Server that allows user to get events streamed from a task

### PR DESCRIPTION
**Pull request by nerdai**
Adds the following endpoints to `deployments` router of API Server:

- `post("/{deployment_name}/tasks/create_nowait")` -> returns session id and task id
- `get("/{deployment_name}/events")` -> returns SSE of event.model_dump()
- `get("/{deployment_name}/results")` -> returns result string of task
- `get("/{deployment_name}/sessions")` -> returns list of active sessions
- `post("/{deployment_name}/sessions/delete")` -> deletes an active session

Where applicable, I've used `query params` instead of path params. If we prefer the latter, I can make a quick fix.

### CODE REVIEW

File: llama_deploy/apiserver/routers/deployments.py [hunk 1]
```diff
@@ -61,6 +63,68 @@ async def create_deployment_task(
     return JSONResponse(result)
 
 
+@deployments_router.post("/{deployment_name}/tasks/create_nowait")
+async def create_deployment_task_nowait(
```
Comment (masci) [inline llama_deploy/apiserver/routers/deployments.py:81]: What if we make this the default (e.g. `@deployments_router.post("/{deployment_name}/tasks/create")`) and we change `create_deployment_task()` to respond to something like `@deployments_router.post("/{deployment_name}/tasks/run")`?

File: llama_deploy/apiserver/routers/deployments.py [hunk 2]
```diff
@@ -61,6 +63,68 @@ async def create_deployment_task(
...
+
+    session = await deployment.client.create_session()
+    task_id = await session.run_nowait(
+        task_definition.agent_id or "", **json.loads(task_definition.input)
+    )
+
+    return JSONResponse({"session_id": session.session_id, "task_id": task_id})
+
+
+@deployments_router.get("/{deployment_name}/events")
```
Comment (masci) [inline llama_deploy/apiserver/routers/deployments.py]: nit: I wonder if `@deployments_router.get("/{deployment_name}/{task_id}/events")` and passing `session_id` as url param is more consistent with the other endpoints?

**Review by masci — COMMENTED** [review]
Very cool this only takes this little amount of code! Left a couple of comments regarding api design, open to discussion!

File: llama_deploy/apiserver/routers/deployments.py (hunk 1 above)
Comment (nerdai) [reply llama_deploy/apiserver/routers/deployments.py:81]: I like this: "run" would make me feel like I should have a result come back at the other side of my request.

File: llama_deploy/apiserver/routers/deployments.py (hunk 2 above)
Comment (nerdai) [reply llama_deploy/apiserver/routers/deployments.py]: Yeah, its a good point. I just saw in our clients we opt for path params.

I was referring to [this](https://stackoverflow.com/a/31261026/19302210) to determine when to use one over the other, but I view consistency as being super important. So, I'd be good with going with path params.

Comment (nerdai) [reply llama_deploy/apiserver/routers/deployments.py]: Sorry, for some reason I don't know how to read! I don't think I've seen us use `session_id` as an url param in our other endpoints?

As mentioned above tho, we have examples in our clients that use session_id and task_id as path params. See here, our control plane server: https://github.com/run-llama/llama_deploy/blob/5e7703e98faa8d682a679832872094258a172629/llama_deploy/control_plane/server.py#L218

Comment (nerdai) [reply llama_deploy/apiserver/routers/deployments.py]: I think doing the above is confusing because we already have `("/{deployment_name}/tasks/create")`, which doesn't require `session_id`. I'm guessing this is why you suggested `session_id` as a url param -- i'd be good with this.

Comment (nerdai) [reply llama_deploy/apiserver/routers/deployments.py]: Alright, so i've modified the routes that require both `task_id` and `session_id` as follows

- `get("/deployments/{deployment_name}/tasks/{task_id}/events")` -> returns SSE of event.model_dump()
- `get("/deployments/{deployment_name}/tasks/{task_id}/results")` -> returns result string of task

Here, `session_id` is passed in as an url param.

I think this works better since we also have these endpoints dedicated to `sessions` resource:

- `get("/deployments/{deployment_name}/sessions")` -> returns list of active sessions
- `post("/deployments/{deployment_name}/sessions/delete")` -> deletes an active session

File: llama_deploy/apiserver/routers/deployments.py [hunk 3]
```diff
@@ -61,15 +77,87 @@ async def create_deployment_task(
...
+    task_id = await session.run_nowait(
+        task_definition.agent_id or "", **json.loads(task_definition.input)
+    )
+
+    return JSONResponse({"session_id": session.session_id, "task_id": task_id})
+
+
+@deployments_router.get("/{deployment_name}/tasks/{task_id}/events")
+async def get_events(
+    deployment_name: str, session_id: str, task_id: str
```
Comment (logan-markewich) [inline llama_deploy/apiserver/routers/deployments.py:107]: session_id is in the args, but not in the URL -- should it be in the url? `/{deployment_name}/{session_id}/tasks/{task_id}/events` ?

File: llama_deploy/apiserver/routers/deployments.py [hunk 4]
```diff
@@ -61,15 +77,87 @@ async def create_deployment_task(
...
+
+    return StreamingResponse(
+        event_stream(),
+        media_type="application/x-ndjson",
     )
+
+
+@deployments_router.get("/{deployment_name}/tasks/{task_id}/results")
+async def get_task_result(
+    deployment_name: str, session_id: str, task_id: str
```
Comment (logan-markewich) [inline llama_deploy/apiserver/routers/deployments.py:129]: similar comment about session_id here

**Review by logan-markewich — APPROVED** [review]
Nice! Also yay for tests

Just one comment about the session_id in a few endpoints, but looks good otherwise

File: llama_deploy/apiserver/routers/deployments.py (hunk 4 above)
Comment (nerdai) [reply llama_deploy/apiserver/routers/deployments.py:129]: hey hey,

yeah there's some discussion related to this above with massi -- it's a bit long and so probably hard to parse 😅

https://github.com/run-llama/llama_deploy/pull/301#discussion_r1794260210

File: llama_deploy/apiserver/routers/deployments.py (hunk 3 above)
Comment (nerdai) [reply llama_deploy/apiserver/routers/deployments.py:107]: hey hey,

yeah there's some discussion related to this above with massi -- it's a bit long and so probably hard to parse 😅

https://github.com/run-llama/llama_deploy/pull/301#discussion_r1794260210

The jist of it is that there could be some good reason to pass in `session_id` as an url/query param vs a path param as you've specified. You can read the linked thread for that rationale. Please let me know what you think.

Comment (logan-markewich) [reply llama_deploy/apiserver/routers/deployments.py:107]: Ah yea, i saw that. I guess it's just been a hot minute since I used query params in fastapi, I guess this is indeed how they work lol

I agree that if the apiserver is using session_id as a query param, let's keep doing that

**Review by masci — APPROVED** [review]
Noice!
